File: app/src/main/java/demo/simpleauth/MainActivity.java
package demo.simpleauth;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.Toast;

import com.jaychang.sa.AuthCallback;
import com.jaychang.sa.SimpleAuth;
import com.jaychang.sa.SocialUser;

import java.util.Arrays;
import java.util.List;

import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainActivity extends AppCompatActivity {

  public static final String FACEBOOK = "FACEBOOK";
  public static final String GOOGLE = "GOOGLE";
  public static final String TWITTER = "TWITTER";
  public static final String INSTAGRAM = "INSTAGRAM";

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    ButterKnife.bind(this);
  }

  @OnClick(R.id.connectFbButton)
  void connectFacebook() {
    List<String> scopes = Arrays.asList("user_birthday", "user_friends");

    SimpleAuth.getInstance().connectFacebook(scopes, new AuthCallback() {
      @Override
      public void onSuccess(SocialUser socialUser) {
        ProfileActivity.start(MainActivity.this, FACEBOOK, socialUser);
      }

      @Override
      public void onError(Throwable error) {
        toast(error.getMessage());
      }

      @Override
      public void onCancel() {
        toast("Canceled");
      }
    });
  }

  @OnClick(R.id.connectGoogleButton)
  void connectGoogle() {
    List<String> scopes = Arrays.asList(
      "https://www.googleapis.com/auth/youtube",
      "https://www.googleapis.com/auth/youtube.upload"
    );

    SimpleAuth.getInstance().connectGoogle(scopes, new AuthCallback() {
      @Override
      public void onSuccess(SocialUser socialUser) {
        ProfileActivity.start(MainActivity.this, GOOGLE, socialUser);
      }

      @Override
      public void onError(Throwable error) {
        toast(error.getMessage());
      }

      @Override
      public void onCancel() {
        toast("Canceled");
      }
    });
  }

  @OnClick(R.id.connectTwitterButton)
  void connectTwitter() {
    SimpleAuth.getInstance().connectTwitter(new AuthCallback() {
      @Override
      public void onSuccess(SocialUser socialUser) {
        ProfileActivity.start(MainActivity.this, TWITTER, socialUser);
      }

      @Override
      public void onError(Throwable error) {
        toast(error.getMessage());
      }

      @Override
      public void onCancel() {
        toast("Canceled");
      }
    });
  }

  @OnClick(R.id.connectIgButton)
  void connectInstagram() {
    List<String> scopes = Arrays.asList("follower_list", "likes");

    SimpleAuth.getInstance().connectInstagram(scopes, new AuthCallback() {
      @Override
      public void onSuccess(SocialUser socialUser) {
        ProfileActivity.start(MainActivity.this, INSTAGRAM, socialUser);
      }

      @Override
      public void onError(Throwable error) {
        toast(error.getMessage());
      }

      @Override
      public void onCancel() {
        toast("Canceled");
      }
    });
  }

  private void toast(String msg) {
    Toast.makeText(MainActivity.this, msg, Toast.LENGTH_LONG).show();
  }

}

File: app/src/main/java/demo/simpleauth/MainActivity.kt
package demo.simpleauth

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.Toast
import com.jaychang.sa.AuthCallback
import com.jaychang.sa.SimpleAuth
import com.jaychang.sa.SocialUser
import com.jaychang.sa.facebook.connectFacebook
import com.jaychang.sa.google.connectGoogle
import com.jaychang.sa.instagram.connectInstagram
import com.jaychang.sa.twitter.connectTwitter
import java.util.*

class MainActivity : AppCompatActivity() {

  companion object {
    const val FACEBOOK = "FACEBOOK"
    const val GOOGLE = "GOOGLE"
    const val TWITTER = "TWITTER"
    const val INSTAGRAM = "INSTAGRAM"
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
  }

  fun connectFacebook(view: View) {
    val scopes = Arrays.asList("user_birthday", "user_friends")

    SimpleAuth.connectFacebook(scopes, object : AuthCallback {
      override fun onSuccess(socialUser: SocialUser) {
        ProfileActivity.start(this@MainActivity, FACEBOOK, socialUser)
      }

      override fun onError(error: Throwable) {
        toast(error.message ?: "")
      }

      override fun onCancel() {
        toast("Canceled")
      }
    })
  }

  fun connectGoogle(view: View) {
    val scopes = Arrays.asList(
      "https://www.googleapis.com/auth/youtube",
      "https://www.googleapis.com/auth/youtube.upload"
    )

    SimpleAuth.connectGoogle(scopes, object : AuthCallback {
      override fun onSuccess(socialUser: SocialUser) {
        ProfileActivity.start(this@MainActivity, GOOGLE, socialUser)
      }

      override fun onError(error: Throwable) {
        toast(error.message ?: "")
      }

      override fun onCancel() {
        toast("Canceled")
      }
    })
  }

  fun connectTwitter(view: View) {
    SimpleAuth.connectTwitter(object : AuthCallback {
      override fun onSuccess(socialUser: SocialUser) {
        ProfileActivity.start(this@MainActivity, TWITTER, socialUser)
      }

      override fun onError(error: Throwable) {
        toast(error.message ?: "")
      }

      override fun onCancel() {
        toast("Canceled")
      }
    })
  }

  fun connectInstagram(view: View) {
    val scopes = Arrays.asList("follower_list", "likes")

    SimpleAuth.connectInstagram(scopes, object : AuthCallback {
      override fun onSuccess(socialUser: SocialUser) {
        ProfileActivity.start(this@MainActivity, INSTAGRAM, socialUser)
      }

      override fun onError(error: Throwable) {
        toast(error.message ?: "")
      }

      override fun onCancel() {
        toast("Canceled")
      }
    })
  }

  private fun toast(msg: String) {
    Toast.makeText(this@MainActivity, msg, Toast.LENGTH_LONG).show()
  }
}
